### PR TITLE
rpc: sort listdescriptors result

### PR DESCRIPTION
**Pull request by Sjors**
This puts receive and change descriptors directly below each other.

The change would be simpler if `UniValue` arrays were sortable.

### CODE REVIEW

**Comment by theStack** [comment]
Concept ACK

Just a few days ago while playing around with descriptor wallet stuff I thought that a sorted output order for `listdescriptors` would be more convenient 👍

**Comment by brunoerg** [comment]
Concept ACK

**Review by w0xlt — APPROVED** [review]
Maybe there could be a well-defined ordering for descriptors (instead of string ordering), but this solution is better than the current code.

ACK https://github.com/bitcoin/bitcoin/pull/25931/commits/4996ae3affa3a3b39c5ec53e74c9e4435b9b871a

**Comment by kristapsk** [comment]
Concept ACK

**Comment by S3RK** [comment]
Concept ACK

File: src/wallet/rpc/backup.cpp [hunk 1]
```diff
@@ -1784,25 +1784,40 @@ RPCHelpMan listdescriptors()
 
     LOCK(wallet->cs_wallet);
 
-    UniValue descriptors(UniValue::VARR);
     const auto active_spk_mans = wallet->GetActiveScriptPubKeyMans();
+
+    std::vector<std::tuple<std::string, WalletDescriptor, bool, std::optional<bool>>> wallet_descriptors;
```
Comment (S3RK) [inline src/wallet/rpc/backup.cpp:1789]: nit: I find tuples a bit hard to grok. Could we just pass a comparator to `GetAllScriptPubKeyMans`?

**Comment by Sjors** [comment]
@S3RK I refactored the PR to use a `struct` (with trivial types) instead of a tuple. This should also make it easier for a future refactor to move some of this stuff out of the RPC and into the wallet.

**Comment by S3RK** [comment]
Code review ACK edbd1aa43f940910f31037770e386e4cf48ab92f

File: src/wallet/rpc/backup.cpp [hunk 2]
```diff
@@ -1784,34 +1784,59 @@ RPCHelpMan listdescriptors()
...
     LOCK(wallet->cs_wallet);
 
-    UniValue descriptors(UniValue::VARR);
     const auto active_spk_mans = wallet->GetActiveScriptPubKeyMans();
+
+    struct WalletDescInfo {
+        std::string descriptor;
+        uint64_t creation_time;
+        bool active;
+        std::optional<bool> internal;
```
Comment (furszy) [inline src/wallet/rpc/backup.cpp:1793]: this field doesn't need to be an optional. It will always have a value.

File: src/wallet/rpc/backup.cpp [hunk 3]
```diff
@@ -1784,34 +1784,59 @@ RPCHelpMan listdescriptors()
...
-        if (internal.has_value()) {
-            spk.pushKV("internal", *internal);
+        const bool is_range = wallet_descriptor.descriptor->IsRange();
+        wallet_descriptors.push_back({
+            descriptor,
+            wallet_descriptor.creation_time,
+            active_spk_mans.count(desc_spk_man) != 0,
+            wallet->IsInternalScriptPubKeyMan(desc_spk_man),
+            is_range ? std::optional(std::make_pair(wallet_descriptor.range_start, wallet_descriptor.range_end)) : std::nullopt,
+            is_range ? wallet_descriptor.next_index : 0
```
Comment (furszy) [inline src/wallet/rpc/backup.cpp:1817]: I don't think that this is needed. if descriptor is ranged, `next_index` should always be 0.

File: src/wallet/rpc/backup.cpp (hunk 2 above)
Comment (S3RK) [reply src/wallet/rpc/backup.cpp:1793]: when a descriptor is not active we can't determine whether it's internal or not

File: src/wallet/rpc/backup.cpp (hunk 3 above)
Comment (S3RK) [reply src/wallet/rpc/backup.cpp:1817]: `next_index` defines the index at which we will derive next address for this descriptor. What should it be always 0?

File: src/wallet/rpc/backup.cpp (hunk 2 above)
Comment (Sjors) [reply src/wallet/rpc/backup.cpp:1793]: Indeed, this is why `IsInternalScriptPubKeyMan` returns `std::optional<bool>`. We list both active and inactive descriptors.

File: src/wallet/rpc/backup.cpp (hunk 3 above)
Comment (Sjors) [reply src/wallet/rpc/backup.cpp:1817]: `next_index` is initialized to zero (see `wallet_util.h`) so I dropped the ternary operator.

Comment (furszy) [reply src/wallet/rpc/backup.cpp]: > next_index defines the index at which we will derive next address for this descriptor. What should it be always 0?

@S3RK I missed two letters there, wanted to say "if descriptor **isn't** ranged".

File: src/wallet/rpc/backup.cpp (hunk 2 above)
Comment (furszy) [reply src/wallet/rpc/backup.cpp:1793]: k gotcha, nvm then.

**Review by furszy — APPROVED** [review]
utACK 50996241

**Comment by Sjors** [comment]
CI failure is fixed in #25963, not worth rebasing.

**Review by w0xlt — APPROVED** [review]
reACK https://github.com/bitcoin/bitcoin/pull/25931/commits/50996241f2b0eefeaab4fcd11b9730fa2dc107ae

**Comment by S3RK** [comment]
reACK 50996241f2b0eefeaab4fcd11b9730fa2dc107ae

**Pull request merged**